ftpvl/processors.py

```diff
@@ -0,0 +1,420 @@
+    Evaluations.
+    """
+
+    def process(self, input_eval: Evaluation) -> Evaluation:
+        return Evaluation(input_eval.get_df() - 1, input_eval.get_eval_id())
+
+
+class StandardizeTypes(Processor):
+    """
+    Processor that casts metrics in an Evaluation to the specified type.
+
+    The type of each metric in an Evaluation is inferred after
+    fetching. This processor accepts a dictionary of types and casts the
+    Evaluation to those types.
+
+    Parameters
+    ----------
+    types : dict
+        A mapping from column names to types
+    """
+
+    def __init__(self, types: dict):
+        self.types = types
+
+    def process(self, input_eval: Evaluation) -> Evaluation:
+        input_df = input_eval.get_df()
+
+        # if int, we might need to convert to float first
+        # (e.g. int(float("6.0")))
+        if int in self.types.values():
+            # create dictionary replacing int with float
+            pre_df_types = {
+                k: (v if v != int else float) for k, v in self.types.items()
+            }
+            input_df = input_df.astype(pre_df_types)
+
+        new_df = input_df.astype(self.types)
+        return Evaluation(new_df, input_eval.get_eval_id())
+
+
+class CleanDuplicates(Processor):
+    """
+    Processor that removes duplicate rows from an Evaluation dataframe based on
+    one or more values in each row.
+
+    By default, the first instance of a duplicate is retained, and all others
+    are removed. You can optionally specify columns to sort by and which way to
+    sort, which provides fine-grained control over which rows are removed.
+
+    Parameters
+    ----------
+    duplicate_col_names : List[str]
+        column names to use when finding duplicates
+
+    sort_col_names : List[str], optional
+        column names to sort by, by default None
+ 
+    reverse_sort : bool, optional
+        sort in ascending order, by default False
+    """
+
+    def __init__(
+        self,
+        duplicate_col_names: List[str],
+        sort_col_names: List[str] = None,
+        reverse_sort: bool = False,
+    ):
+
+        self._duplicate_col_names = duplicate_col_names
+        self._sort_col_names = sort_col_names
+        self._reverse_sort = reverse_sort
+
+    def process(self, input_eval: Evaluation) -> Evaluation:
+        if self._sort_col_names is None:
+            new_df = input_eval.get_df().drop_duplicates(
+                subset=self._duplicate_col_names
+            )
+            return Evaluation(new_df, input_eval.get_eval_id())
+        else:
+            new_df = (
+                input_eval.get_df()
+                .sort_values(by=self._sort_col_names, ascending=self._reverse_sort)
+                .drop_duplicates(subset=self._duplicate_col_names)
+            )
+            return Evaluation(new_df, input_eval.get_eval_id())
+
+
+class AddNormalizedColumn(Processor):
+    """
+    Processor that groups rows by a column, calculates the maximum of the
```

Comment:
For runtime, it makes more sense to normalize based on the minimum. Generally the comparison is made to the "best" measurement.

Reply:
Added support for direction to AddNormalizedColumn. Can specify to use either min or max for the baseline when normalizing.